source/agora/test/Byzantine.d

```diff
@@ -0,0 +1,135 @@
+        TaskManager taskman)
+    {
+        super(params, clock, network, key_pair, ledger, taskman,
+            this.txs_to_nominate);
+    }
+
+    // refuse to sign
+    extern(C++) override void signEnvelope (ref SCPEnvelope envelope)
+    {
+    }
+}
+
+/// node which refuses to co-operate: doesn't sign / forges the signature / etc
+class ByzantineNode : TestValidatorNode
+{
+    public this (Config config, Registry* reg, immutable(Block)[] blocks,
+        ulong txs_to_nominate, shared(time_t)* cur_time)
+    {
+        super(config, reg, blocks, txs_to_nominate, cur_time);
+    }
+
+    protected override TestNominator getNominator (immutable(ConsensusParams) params,
+        Clock clock, NetworkManager network, KeyPair key_pair, Ledger ledger,
+        TaskManager taskman)
+    {
+        return new BadNominator(params, clock, network, key_pair, ledger,
+            taskman);
+    }
+}
+
+/// creates `bad_count` nodes which will refuse to sign
+class ByzantineManager (size_t bad_count) : TestAPIManager
+{
+    ///
+    public this (immutable(Block)[] blocks, TestConf test_conf,
+        time_t genesis_start_time)
+    {
+        super(blocks, test_conf, genesis_start_time);
+    }
+
+    public override void createNewNode (Config conf,
+        string file = __FILE__, int line = __LINE__)
+    {
+        if (this.nodes.length < bad_count)
+        {
+            auto time = new shared(time_t)(this.initial_time);
+            assert(conf.node.is_validator);
+            auto node = RemoteAPI!TestAPI.spawn!ByzantineNode(
+                conf, &this.reg, this.blocks, this.test_conf.txs_to_nominate,
+                time, conf.node.timeout);
+            this.reg.register(conf.node.address, node.ctrl.tid());
+            this.nodes ~= NodePair(conf.node.address, node, time);
+        }
+        else
+            super.createNewNode(conf, file, line);
+    }
+}
+
+///
+unittest
+{
+    TestConf conf = { validators : 4, quorum_threshold : 66 };
+    auto network = makeTestNetwork!(ByzantineManager!1)(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+    auto nodes = network.clients;
+    auto node_1 = nodes[$ - 1];
+    auto txes = genesisSpendable().map!(txb => txb.sign()).array();
+    txes.each!(tx => node_1.putTransaction(tx));
+    network.expectBlock(Height(1), 5.seconds);
+}
+
+///
+unittest
+{
+    TestConf conf = { validators : 4, quorum_threshold : 66 };
+    auto network = makeTestNetwork!(ByzantineManager!2)(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+    auto nodes = network.clients;
+    auto node_1 = nodes[$ - 1];
+    auto txes = genesisSpendable().map!(txb => txb.sign()).array();
+    txes.each!(tx => node_1.putTransaction(tx));
+    assertThrown!AssertError(network.expectBlock(Height(1), 5.seconds));
```

Comment:
So this test will always take 5 seconds. That's not going to scale for byzantine nodes test. Also testing assert error isn't great. And this is essentially testing that an event didn't happen during a time frame, but it would be much better if we could observe that the bad behavior happened instead. Maybe hook into signEnvelop ?